File: libs/python/stonecharts/charts/boxplot.py
"""Boxplot chart renderer: ChartSpec -> SVG string.

Box-and-whisker glyphs over a categorical axis. Each glyph shows a 5-number
summary (low, q1, median, q3, high) plus optional outlier circles. Shared
Cartesian chrome comes from _cartesian.py; this module draws only the boxplot
marks. See charts/boxplot/design.md for the full geometry contract.
"""

from __future__ import annotations

import copy

from ..spec import ChartSpec
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, render_cartesian

PAD = 0.2
CAP = 0.5
BOX_OPACITY = 0.5
OUTLIER_R = 2.5
MIN_BOX = 1.0


def render_svg(spec: ChartSpec) -> str:
    mod = copy.copy(spec)
    mod.y_axis = copy.copy(spec.y_axis)

    all_vals: list[float] = []
    for s in mod.series:
        for bd in s.box_data or []:
            all_vals.append(bd.low)
            all_vals.append(bd.high)
            all_vals.extend(bd.outliers)

    if not all_vals:
        all_vals = [v for s in mod.series for v in s.data]

    if mod.y_axis.min is None and all_vals:
        mod.y_axis.min = min(all_vals)
    if mod.y_axis.max is None and all_vals:
        mod.y_axis.max = max(all_vals)

    orientation = mod.orientation or "vertical"
    return render_cartesian(mod, "Boxplot", "band", _boxplot_marks, include_zero=False, orientation=orientation)


def _boxplot_marks(fr: CartesianFrame, p: list) -> None:
    if fr.n <= 0:
        return

    spec = fr.spec
    horiz = fr.orientation == "horizontal"
    K = max(len(spec.series), 1)

    for si, s in enumerate(spec.series):
        st = fr.styles[si]
        box_data = s.box_data or []
        p.append(f'<g class="sc-series" data-series="{si}">')

        for i in range(min(len(box_data), fr.n)):
            bd = box_data[i]
            cat = fr.cats[i] if i < len(fr.cats) else str(i)

            if horiz:
                band_sz = fr.band_height()
                group_sz = band_sz * (1 - PAD)
                bar_sz = group_sz / K
                band_c = fr.band_center(i)
                slot_start = band_c - group_sz / 2 + bar_sz * si
                mid = slot_start + bar_sz / 2
                cap_half = bar_sz * CAP / 2

                xq1 = fr.value_pix(bd.q1)
                xq3 = fr.value_pix(bd.q3)
                xmed = fr.value_pix(bd.median)
                xlow = fr.value_pix(bd.low)
                xhigh = fr.value_pix(bd.high)

                bx = min(xq1, xq3)
                bw = abs(xq3 - xq1)
                if bw < MIN_BOX:
                    bw = MIN_BOX

                cx_val = xmed
                cy_val = mid

                common = (
                    f'class="sc-box sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                    f'data-x="{esc(cat)}" data-y="{esc(fmt_num(bd.median))}" '
                    f'data-color="{st.solid}" data-r="3.5" data-r-hover="6"'
                )
                p.append(
                    f'<rect {common} cx="{cx_val:.1f}" cy="{cy_val:.1f}" '
                    f'x="{bx:.1f}" y="{slot_start:.1f}" width="{bw:.1f}" height="{bar_sz:.1f}" '
                    f'fill="{st.fill}" fill-opacity="{BOX_OPACITY}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-median" x1="{xmed:.1f}" y1="{slot_start:.1f}" '
                    f'x2="{xmed:.1f}" y2="{slot_start + bar_sz:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker" x1="{xq3:.1f}" y1="{mid:.1f}" '
                    f'x2="{xhigh:.1f}" y2="{mid:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker-cap" x1="{xhigh:.1f}" y1="{mid - cap_half:.1f}" '
                    f'x2="{xhigh:.1f}" y2="{mid + cap_half:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker" x1="{xq1:.1f}" y1="{mid:.1f}" '
                    f'x2="{xlow:.1f}" y2="{mid:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker-cap" x1="{xlow:.1f}" y1="{mid - cap_half:.1f}" '
                    f'x2="{xlow:.1f}" y2="{mid + cap_half:.1f}" stroke="{st.solid}"/>'
                )
                for o in bd.outliers:
                    ox = fr.value_pix(o)
                    p.append(
                        f'<circle class="sc-outlier" cx="{ox:.1f}" cy="{mid:.1f}" r="{OUTLIER_R}" fill="{st.solid}"/>'
                    )
            else:
                band_sz = fr.band_width()
                group_sz = band_sz * (1 - PAD)
                bar_sz = group_sz / K
                band_c = fr.xpix(i)
                slot_start = band_c - group_sz / 2 + bar_sz * si
                mid = slot_start + bar_sz / 2
                cap_half = bar_sz * CAP / 2

                yq3 = fr.ypix(bd.q3)
                yq1 = fr.ypix(bd.q1)
                ymed = fr.ypix(bd.median)
                yhigh = fr.ypix(bd.high)
                ylow = fr.ypix(bd.low)

                box_h = yq1 - yq3
                if box_h < MIN_BOX:
                    box_h = MIN_BOX

                cx_val = mid
                cy_val = ymed

                common = (
                    f'class="sc-box sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                    f'data-x="{esc(cat)}" data-y="{esc(fmt_num(bd.median))}" '
                    f'data-color="{st.solid}" data-r="3.5" data-r-hover="6"'
                )
                p.append(
                    f'<rect {common} cx="{cx_val:.1f}" cy="{cy_val:.1f}" '
                    f'x="{slot_start:.1f}" y="{yq3:.1f}" width="{bar_sz:.1f}" height="{box_h:.1f}" '
                    f'fill="{st.fill}" fill-opacity="{BOX_OPACITY}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-median" x1="{slot_start:.1f}" y1="{ymed:.1f}" '
                    f'x2="{slot_start + bar_sz:.1f}" y2="{ymed:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker" x1="{mid:.1f}" y1="{yq3:.1f}" '
                    f'x2="{mid:.1f}" y2="{yhigh:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker-cap" x1="{mid - cap_half:.1f}" y1="{yhigh:.1f}" '
                    f'x2="{mid + cap_half:.1f}" y2="{yhigh:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker" x1="{mid:.1f}" y1="{yq1:.1f}" '
                    f'x2="{mid:.1f}" y2="{ylow:.1f}" stroke="{st.solid}"/>'
                )
                p.append(
                    f'<line class="sc-whisker-cap" x1="{mid - cap_half:.1f}" y1="{ylow:.1f}" '
                    f'x2="{mid + cap_half:.1f}" y2="{ylow:.1f}" stroke="{st.solid}"/>'
                )
                for o in bd.outliers:
                    oy = fr.ypix(o)
                    p.append(
                        f'<circle class="sc-outlier" cx="{mid:.1f}" cy="{oy:.1f}" r="{OUTLIER_R}" fill="{st.solid}"/>'
                    )

        p.append("</g>")

File: libs/go/boxplot.go
package stonecharts

import (
	"fmt"
	"math"
	"strings"
)

const boxplotPad = 0.2
const boxplotCap = 0.5
const boxplotOpacity = 0.5
const boxplotOutlierR = 2.5
const boxplotMinBox = 1.0

func renderBoxplotSVG(spec *ChartSpec) string {
	s2 := *spec
	s2.Series = make([]Series, len(spec.Series))
	copy(s2.Series, spec.Series)

	var allVals []float64
	for _, s := range s2.Series {
		for _, bd := range s.BoxData {
			allVals = append(allVals, bd.Low, bd.High)
			allVals = append(allVals, bd.Outliers...)
		}
	}
	if len(allVals) == 0 {
		for _, s := range s2.Series {
			allVals = append(allVals, s.Data...)
		}
	}

	if s2.YAxis.Min == nil && len(allVals) > 0 {
		mn := allVals[0]
		for _, v := range allVals[1:] {
			if v < mn {
				mn = v
			}
		}
		s2.YAxis.Min = &mn
	}
	if s2.YAxis.Max == nil && len(allVals) > 0 {
		mx := allVals[0]
		for _, v := range allVals[1:] {
			if v > mx {
				mx = v
			}
		}
		s2.YAxis.Max = &mx
	}

	orientation := s2.Orientation
	if orientation == "" {
		orientation = "vertical"
	}
	return renderCartesian(&s2, "Boxplot", "band", boxplotMarks, false, orientation)
}

func boxplotMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}
	spec := f.spec
	horiz := f.orientation == "horizontal"
	K := len(spec.Series)
	if K <= 0 {
		K = 1
	}

	for si, s := range spec.Series {
		st := f.styles[si]
		boxData := s.BoxData
		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))

		count := len(boxData)
		if count > f.n {
			count = f.n
		}

		for i := 0; i < count; i++ {
			bd := boxData[i]
			cat := fmt.Sprintf("%d", i)
			if i < len(f.cats) {
				cat = f.cats[i]
			}

			if horiz {
				bandSz := f.bandHeight()
				groupSz := bandSz * (1 - boxplotPad)
				barSz := groupSz / float64(K)
				bandC := f.bandCenter(i)
				slotStart := bandC - groupSz/2 + barSz*float64(si)
				mid := slotStart + barSz/2
				capHalf := barSz * boxplotCap / 2

				xq1 := f.valuePix(bd.Q1)
				xq3 := f.valuePix(bd.Q3)
				xmed := f.valuePix(bd.Median)
				xlow := f.valuePix(bd.Low)
				xhigh := f.valuePix(bd.High)

				bx := math.Min(xq1, xq3)
				bw := math.Abs(xq3 - xq1)
				if bw < boxplotMinBox {
					bw = boxplotMinBox
				}

				cxVal := xmed
				cyVal := mid

				common := fmt.Sprintf(
					`class="sc-box sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="3.5" data-r-hover="6"`,
					si, esc(s.Name), esc(cat), esc(fmtNum(bd.Median)), st.solid)
				p.WriteString(fmt.Sprintf(
					`<rect %s cx="%s" cy="%s" x="%s" y="%s" width="%s" height="%s" fill="%s" fill-opacity="%s" stroke="%s"/>`,
					common, f1(cxVal), f1(cyVal), f1(bx), f1(slotStart), f1(bw), f1(barSz), st.fill, fmtNum(boxplotOpacity), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-median" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(xmed), f1(slotStart), f1(xmed), f1(slotStart+barSz), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(xq3), f1(mid), f1(xhigh), f1(mid), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker-cap" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(xhigh), f1(mid-capHalf), f1(xhigh), f1(mid+capHalf), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(xq1), f1(mid), f1(xlow), f1(mid), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker-cap" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(xlow), f1(mid-capHalf), f1(xlow), f1(mid+capHalf), st.solid))
				for _, o := range bd.Outliers {
					ox := f.valuePix(o)
					p.WriteString(fmt.Sprintf(
						`<circle class="sc-outlier" cx="%s" cy="%s" r="%s" fill="%s"/>`,
						f1(ox), f1(mid), fmtNum(boxplotOutlierR), st.solid))
				}
			} else {
				bandSz := f.bandWidth()
				groupSz := bandSz * (1 - boxplotPad)
				barSz := groupSz / float64(K)
				bandC := f.xpix(float64(i))
				slotStart := bandC - groupSz/2 + barSz*float64(si)
				mid := slotStart + barSz/2
				capHalf := barSz * boxplotCap / 2

				yq3 := f.ypix(bd.Q3)
				yq1 := f.ypix(bd.Q1)
				ymed := f.ypix(bd.Median)
				yhigh := f.ypix(bd.High)
				ylow := f.ypix(bd.Low)

				boxH := yq1 - yq3
				if boxH < boxplotMinBox {
					boxH = boxplotMinBox
				}

				cxVal := mid
				cyVal := ymed

				common := fmt.Sprintf(
					`class="sc-box sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="3.5" data-r-hover="6"`,
					si, esc(s.Name), esc(cat), esc(fmtNum(bd.Median)), st.solid)
				p.WriteString(fmt.Sprintf(
					`<rect %s cx="%s" cy="%s" x="%s" y="%s" width="%s" height="%s" fill="%s" fill-opacity="%s" stroke="%s"/>`,
					common, f1(cxVal), f1(cyVal), f1(slotStart), f1(yq3), f1(barSz), f1(boxH), st.fill, fmtNum(boxplotOpacity), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-median" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(slotStart), f1(ymed), f1(slotStart+barSz), f1(ymed), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(mid), f1(yq3), f1(mid), f1(yhigh), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker-cap" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(mid-capHalf), f1(yhigh), f1(mid+capHalf), f1(yhigh), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(mid), f1(yq1), f1(mid), f1(ylow), st.solid))
				p.WriteString(fmt.Sprintf(
					`<line class="sc-whisker-cap" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s"/>`,
					f1(mid-capHalf), f1(ylow), f1(mid+capHalf), f1(ylow), st.solid))
				for _, o := range bd.Outliers {
					oy := f.ypix(o)
					p.WriteString(fmt.Sprintf(
						`<circle class="sc-outlier" cx="%s" cy="%s" r="%s" fill="%s"/>`,
						f1(mid), f1(oy), fmtNum(boxplotOutlierR), st.solid))
				}
			}
		}

		p.WriteString(`</g>`)
	}
}
